crates/modelardb_server/src/storage/types.rs

```diff
@@ -75,6 +85,45 @@ impl MemoryPool {
         }
     }
 
+    /// Change the amount of memory available for multivariate data by `size_in_bytes`.
+    pub(super) fn adjust_multivariate_memory(&self, size_in_bytes: isize) {
+        // unwrap() is safe as lock() only returns an error if the mutex is poisoned.
+        *self.remaining_multivariate_memory_in_bytes.lock().unwrap() += size_in_bytes;
+        self.wait_for_multivariate_memory.notify_all();
+    }
+
+    /// Return the amount of memory available for multivariate data in bytes.
+    #[cfg(test)]
+    #[must_use]
+    pub(super) fn remaining_multivariate_memory_in_bytes(&self) -> isize {
+        // unwrap() is safe as lock() only returns an error if the mutex is poisoned.
+        *self.remaining_multivariate_memory_in_bytes.lock().unwrap()
+    }
+
+    /// Wait until `size_in_bytes` bytes of memory is available for multivariate data and then
+    /// reserve it.
+    pub(super) fn wait_for_multivariate_memory(&self, size_in_bytes: usize) {
+        // unwrap() is safe as lock() only returns an error if the mutex is poisoned.
+        let mut memory_in_bytes = self.remaining_multivariate_memory_in_bytes.lock().unwrap();
+
+        while *memory_in_bytes < size_in_bytes as isize {
+            // unwrap() is safe as wait() only returns an error if the mutex is poisoned.
+            memory_in_bytes = self
+                .wait_for_multivariate_memory
+                .wait(memory_in_bytes)
+                .unwrap();
+        }
+
+        *memory_in_bytes -= size_in_bytes as isize;
+    }
+
+    /// Free `size_in_bytes` bytes of memory for storing multivariate data.
+    pub(super) fn free_multivariate_memory(&self, size_in_bytes: usize) {
```

Comment:
The methods differ in name, documentation, and type for its parameter. I kept both because they are very short and the difference in types prevents a negative amount of memory being freed. I agree that if the functions grow in size, the benefits are not worth maintaining both functions. Instead the `free_` method should call the other or be deleted.

Reply:
Would it not be better then to have a `reserve_` and `free_` function for each and remove `adjust_`? Right now it seems we have one function that can be used for both reserving and freeing memory (`adjust`) and one that can be used for freeing.

Reply:
Since this will require changes to `Configuration` and this PR is already very big, I have created a task to do this with a link to this discussion.